### ISSUE TITLE
LLVM 3.6 support on 0.3?

### ISSUE THREAD

**Issue by nalimilan** [JuliaLang/julia#10786]
Is there any chance Julia 0.3 could support LLVM 3.6 with a few backports? Fedora rawhide just switched to it, and the build fails due to a missing header (of course that's only the first step):
```
g++ -march=x86-64 -m64 -I/usr/include -DNDEBUG -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -O3 -fomit-frame-pointer -std=c++11 -fvisibility-inlines-hidden -fno-exceptions -fPIC -ffunction-sections -fdata-sections -Wcast-qual -DJL_SYSTEM_IMAGE_PATH="../lib64/julia/sys.ji" -DJULIA_TARGET_ARCH=x86-64 -DSYSTEM_LLVM -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -pipe -fPIC -fno-rtti -O3 -ggdb3 -falign-functions -momit-leaf-frame-pointer -D_GNU_SOURCE -Iflisp -Isupport -I/usr/include -I/builddir/build/BUILD/julia-0.3.7/deps/libuv/include -I/builddir/build/BUILD/julia-0.3.7/build/usr/include -DLIBRARY_EXPORTS -Wall -Wno-strict-aliasing -fno-omit-frame-pointer -fvisibility=hidden -fno-common -c codegen.cpp -o codegen.o
codegen.cpp:27:51: fatal error: llvm/ExecutionEngine/JITMemoryManager.h: No such file or directory
compilation terminated.
Makefile:56: recipe for target 'codegen.o' failed
make[2]: *** [codegen.o] Error 1
```

That's not a big deal since 0.4 should be out much before rawhide becomes the next Fedora is released. I could ship a 0.4 pre-release in the meantime.

**Comment by tkelman** [comment]
If someone prepares a PR against release-0.3 and it doesn't break anything, I don't see why not. No idea which backports would be necessary, and most likely they won't apply as straight cherry-picks since the surrounding code has gone through many other changes.

**Comment by nalimilan** [comment]
My interrogation was indeed about whether it would be easy to backport these commits or whether they are too mixed with unrelated changes.

**Comment by vtjnash** [comment]
it's unlikely to be easy to identify the changes to backport. it may be possible to walk through the errors one at a time and resolve them. but llvm-3.3 has always been the only officially supported version for Julia-v0.3, due to MCJIT bugs found in the transitionary period since then. it our recommendation that distributions provide all versions of llvm to build against as dependencies, due to the highly significant API changes that occur between each release version of llvm.

**Comment by tkelman** [comment]
Agreed, distributions really need to revisit their practices when it comes to the lifetime of packaging LLVM versions. Continuing to distribute a few additional already-packaged versions is less maintenance work in the end than every single downstream consumer of LLVM having to hack and possibly destabilize their release branches to accommodate switching to a much newer version on a whim based on distributions' release schedules which are completely out of synch from either LLVM or the downstream projects' schedules.

**Comment by JeffBezanson** [comment]
Currently we only support LLVM 3.3 officially.

I also agree that distributions need to change their policy on LLVM versions. For example, soon LLVM will remove the "legacy JIT" entirely. While we plan to move to the new JIT eventually, some projects might depend on the legacy JIT. It would be impossible to use those packages on fedora if older LLVM versions are "banned". Ah, the friendly world of free software, where apparently you're not allowed to use the version of something that you need.

**Comment by nalimilan** [comment]
That's fine, I'll ship a 0.4 prerelease. I much prefer you spend time improving 0.4!

I agree the current practice in Fedora is quite demanding for projects relying on LLVM, and we're always at risk of not even being able to ship Julia in a newer release of Fedora if the release schedules don't match. I had prepared an LLVM 3.3 parallel-installable package, but the feedback was negative at the time -- though from only one person [1]. I think it would be accepted if the LLVM a Fedora version uses was completely incompatible with Julia, as suddenly removing a package is quite a bad practice.

1: https://bugzilla.redhat.com/show_bug.cgi?id=1109390

**Comment by MaheshWaidande** [comment]
I am working on building/porting Julia V0.3.9 source on ppc64le architecture. I am using Ubuntu 14.10 on top of ppc64le hardware. While compiling Julia V0.3.9 with LLVM-3.6.0, I came across with ‘llvm/ExecutionEngine/JITMemoryManager.h: No such file or directory’ error. 

May be above error will got resolved with latest stable release of Julia, I want to give a try with prerelease version V0.4 of Julia, @nalimilan from where I get the pre-release version of Julia? I try to search it on github (https://github.com/JuliaLang/julia ) but I did not find it.
